### PR TITLE
docs: add info for developers of finished products

### PR DESCRIPTION
**Pull request by oclyke**
Hi all,

in #7440 @dpgeorge suggested adding some documentation about best practices for providing production-ready main program execution. I've taken a stab at doing that and I'm hoping to get some feedback.

I think a 'Best Practices' chapter in docs would be useful to people who are brand-new to the codebase or returning after some time. It would also help spread the expert's knowledge without making them repeat themselves as often.

Some particular items we could discuss:
* is it worthwhile to create a brand new top level chapter called 'Best Practices?'
* when talking about freezing code into firmware images I am only familiar with the ESP32 procedure (using manifest.py). I looked around in the other microcontroller ports but did not see the same mechanism. Can we go into detail describing exactly how to freeze code? If so where?
* @dpgeorge said that ```sys.path``` could be used to allow or disallow overriding of duplicate modules (frozen vs. on filesystem) though in my brief search I could not see how the frozen modules were located on ```sys.path```. For this PR I did a quick test verifying that frozen modules are preferred over the filesystem *by default* in the ESP32 port. Can anyone offer insight on this subject?

Thanks for looking!

### CODE REVIEW

**Comment by pidou46** [comment]
Hi,

Thanks for your contribution.

Regarding the location, I would say a subsection in :
"Distribution packages, package management, and deploying applications"

Just my opinion.

**Comment by oclyke** [comment]
Great, I like it! I'll wait a bit for any other comments before making the change

**Comment by oclyke** [comment]
@pidou46 I've just come back to make the change you suggested and had trouble finding a section for "Distribution packages, package management, and deploying applications." Did you mean to create a new category with that title?

**Comment by kadamski** [comment]
@oclyke http://docs.micropython.org/en/latest/reference/packages.html

**Comment by oclyke** [comment]
Thanks, moved under that section with a little intro blurb and TOC tree.

Phew. There's a lot of documentation to read through. I admit I'd never seen that page before! Probably cause I usually am relying on the search engine to get me the docs I need, with only a little time to peruse all the other content. 

Makes me wonder what techniques could be used to get people toward the right docs. Is is simply a matter of taking time to read it all or could an interactive guided application ask the right questions to get someone there? Well, I'm not going to take this off topic any further but that might be a neat tool one day.

**Comment by oclyke** [comment]
Is anyone interested in this addition? If so are there any changes I can make to help get it merged?

**Comment by stinos** [comment]
To me, the first bit about board definitions and adding boot/main.py reads like general documentation (which might be available elsewhere already, no?) not 'best practices', and the rest about protecting applications looks like some _very_ specific guidelines for a certain situation in a certain production scenario you encountered, and it might not be generally applicable. So while you have interesting information in there, my main gripe would be calling this 'best practices'.

**Comment by oclyke** [comment]
Happy to change the presentation of this info. Modified the title of the PR to reflect this - this is just information for developers who may be curious about these items. I had to find these things out by asking around, experimentation, and generally bothering people. The hope is to make this info available to others in a place that is easy to find.

Anyone, please feel free to request or make additional changes.

File: docs/reference/production.rst [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+*protecting a main application*
+
+Guaranteeing proper functionality on a production product is crucial. It can be concerning that users
+have the potential to access the filesystem and create **main.py** / **boot.py**, thereby influencing
+runtime execution. There are two main concerns: removal / destruction of the intended main program and 
+overriding the main program.
+
+* modules frozen into the firmware are write-protected (barring extreme cases such as self-modifying code)
+* when a module exists in both frozen bytecode and on the filesystem the frozen module is preferred
```
Comment (karlp) [inline docs/reference/production.rst:56]: this doesn't seem to be correct.  If I use mpremote to copy a new instance of a module to the filesystem, the new module is chosen over the frozen older version....

Comment (andrewleech) [reply docs/reference/production.rst:56]: it depends on `sys.path`, if `.frozen` is first in that list then the frozen code is preferred over anything on the filesystem. The default `sys.path` might be different on some ports though.... but can always be enforced to a "safe" value in `boot.py`

**Pull request closed without merging**